### ISSUE TITLE
Order Condition Formula on a Discount ignoring date field

### ISSUE THREAD

**Issue by simple-nicholas** [craftcms/commerce#2505]
### Description
Order Condition Formula on a Discount is ignored under unclear circumstances (the Discount saves with no errors). I can't find any indications in any logs or anywhere what the issue is.

In this case, the formula is `order.eventDate < date('1-10-2021')`, where eventDate is naturally a date field. This syntax is apparently correct from everything I can find.

### Steps to reproduce

1. Add a custom date field to Orders in Commerce Settings
2. Create a Discount with an Order Condition Formula evaluating the date field
3. In the front end, add the discount to the cart with a date that violates the Order Condition Formula

### Additional info

- Craft CMS version: 3.7.7
- Craft Commerce version: 3.4.1
- PHP version: 7.4
- Database driver & version: MariaDB, 10.5.8

**Comment by pdaleramirez** [comment]
@simple-nicholas We'll be releasing a fix for this issue. It will be possible to use an operator to compare custom date fields. In your case, you'll write your condition formula like this 
```
'2021-10-01' > order.eventDate|date("Y-m-d")
```
You can track the changes here https://github.com/craftcms/commerce/pull/2507

**Comment by lukeholder** [comment]
Thanks this will be in the next release of commerce.

**Comment by simple-nicholas** [comment]
Please reopen as #2507 has not resolved this issue.

Namely, Order Condition Formulas containing a date filter on a custom date field fails syntax validation.

The date filter and custom date fields appear to pass validation separately, but not together (eg the example from @pdaleramirez).

**Comment by pdaleramirez** [comment]
@simple-nicholas Sorry, the correct syntax for the condition should be
```
'2021-10-01'|date > order.eventDate|date("Y-m-d")
```
You need to add the `|date` filter

**Comment by simple-nicholas** [comment]
@pdaleramirez Your example also doesn't work, I'm still seeing a syntax validation error.

The issue appears to only occur specifically when the date filter is applied to a custom date field (eg removing the second date filter from your latest example passes validation).

Also, does the date filter work differently here? Your latest example uses an empty date filter (`|date`), which normally only results in a date string in the default format (in this case, "1 Oct 2021").

**Comment by lukeholder** [comment]
Reopening and investigating.

**Comment by simple-nicholas** [comment]
Hi, I've been asked to confirm that this issue is still being worked on / hasn't fallen through the cracks?

**Comment by nfourtythree** [comment]
Hi @simple-nicholas 

Thank you for your message.

I have just tested this with the latest version of Craft and Commerce and everything seems to be working as expected.

I created a custom date field, `myDateField`, and added it to the order field layout. Then, I added a discount with a order condition formula as follows:

```twig
order.myDateField and '2022-01-30' > order.myDateField|date('Y-m-d')
```

This formula states that if the custom date field is filled out and has a date that is before  `2022-01-30` the condition will pass and apply the discount.

Here is a video of the discount working as expected:

https://user-images.githubusercontent.com/266453/149327989-fb8dab1e-96df-4bda-999b-45b3edc297d0.mp4

Please let me know if you have specific steps to be able to reproduce this issue, thanks!

**Comment by simple-nicholas** [comment]
Hi @nfourtythree

Unfortunately, the issue is still not resolved.

I tried what you showed in a fresh Craft 3 instance, and saw it working, but the error persisted in an existing site (despite updating everything, clearing caches, etc).

On further investigation, I found we'd both accidentally used the same bad workaround to successfully save the Order Condition Formula with a date filter. If in that video if you'd tried saving the Discount again at the end, it would have failed.

The problem stems from the fact that the condition is validated by testing it against either the first order, or a blank one if none is available. (See the function in commerce/src/models/Discount.php:403).

If the custom date field is blank, it's evaluated as `null`, which the date filter is fine with.

But the validator is receiving the date field as an array, eg `'eventDate' => ['date' => '2018-02-28 00:00:00.000000', 
 timezone_type' => 3, 'timezone' => 'Australia/Adelaide']` (take from a dump of $params in the function at commerce/src/services/Formulas:99) but the date filter is expecting a string. At this point the createTemplate call throws an error (for the date filter expecting a string, but receiving an array), which is reported as a syntax error.

Changing your condition to read `order.eventDate.date and '2022-01-01' > order.eventDate.date|date('Y-m-d')` will cause the validation to pass, but only if there's at least one order and the date field on the first order has a value. So a proper fix is needed.
